File: mcp-code-intelligence-kotlin/src/main/kotlin/com/codeintel/memory/MemoryEngine.kt
/** Facade for the SDLC Memory Engine — single entry point for all memory operations. */
package com.codeintel.memory

import com.codeintel.db.DatabaseManager
import com.codeintel.log
import com.codeintel.memory.db.MemoryDatabaseManager
import com.codeintel.memory.graph.KnowledgeGraph
import com.codeintel.memory.ingest.autolink.*
import com.codeintel.memory.map.EntityRepository
import com.codeintel.memory.repository.*

class MemoryEngine(db: DatabaseManager) {
    private val memoryDb = MemoryDatabaseManager(db)

    val connection: java.sql.Connection get() = memoryDb.conn
    val knowledge = KnowledgeRepository(memoryDb)
    val search = KnowledgeSearchRepository(memoryDb)
    val graph = GraphRepository(memoryDb)
    val vectors = VectorRepository(memoryDb)
    val consolidation = ConsolidationRepository(memoryDb)
    val sessions = SessionRepository(memoryDb)
    val audit = AuditRepository(memoryDb)

    /** AutoLinker instance, available after initialize(). */
    var autoLinker: AutoLinker? = null
        private set

    /** Current active session ID (set after initialize). */
    var currentSessionId: String? = null
        private set

    /** Initialize memory schema and subsystems. */
    fun initialize() {
        memoryDb.initialize()
        autoLinker = createAutoLinker()
        log("Memory engine ready (auto-linker enabled)")
    }

    /** Start a new session and set it as current. */
    fun startSession(agentName: String? = null): String {
        val sid = sessions.startSession(agentName)
        currentSessionId = sid
        audit.log("SESSION_START", sessionId = sid, agentName = agentName)
        log("Session started: $sid")
        return sid
    }

    /** End the current session. */
    fun endSession() {
        val sid = currentSessionId ?: return
        sessions.endSession(sid)
        audit.log("SESSION_END", sessionId = sid)
        currentSessionId = null
        log("Session ended: $sid")
    }

    /** Get overall memory statistics. */
    fun getStats(): MemoryStats {
        val tierStats = consolidation.getTierStats()
        return MemoryStats(
            totalEntries = tierStats.sumOf { it.entryCount },
            totalEdges = graph.countEdges(),
            totalVectors = vectors.count(),
            tierBreakdown = tierStats.associate { it.tier to it.entryCount }
        )
    }

    /** Create AutoLinker with all strategies wired. */
    private fun createAutoLinker(): AutoLinker {
        val knowledgeGraph = KnowledgeGraph(graph).also { it.loadFromDb() }
        val entityRepo = EntityRepository(connection)
        val strategies = listOf<LinkingStrategy>(
            SemanticStrategy(vectors),
            EntityStrategy(entityRepo),
            TagStrategy(connection),
            FtsStrategy(connection)
        )
        return AutoLinker(knowledgeGraph, graph, connection, strategies)
    }
}

/** Summary statistics for the memory system. */
data class MemoryStats(
    val totalEntries: Int,
    val totalEdges: Int,
    val totalVectors: Int,
    val tierBreakdown: Map<String, Int>
)

File: mcp-code-intelligence-nodejs/src/memory/memory-engine.ts
/**
 * MemoryEngine — facade for the SDLC Memory system.
 * Single entry point for all memory operations.
 */

import Database from 'better-sqlite3';
import { MemoryDatabaseManager } from './memory-db.js';
import { KnowledgeRepository } from './knowledge-repo.js';
import { KnowledgeSearchRepository } from './search-repo.js';
import { GraphRepository } from './graph-repo.js';
import { VectorRepository } from './vector-repo.js';
import { ConsolidationRepository } from './consolidation-repo.js';
import { SessionRepository } from './session-repo.js';
import { AuditRepository } from './audit-repo.js';
import { KnowledgeGraph } from './knowledge-graph.js';
import { CoreMemoryManager } from './core-memory.js';
import { EntityRepository } from './entity-repo.js';
import { ConversationRepository } from './conversation-repo.js';
import { TierStats } from './models.js';

export interface MemoryStats {
  totalEntries: number;
  totalEdges: number;
  totalVectors: number;
  tierBreakdown: TierStats[];
}

export class MemoryEngine {
  readonly knowledge: KnowledgeRepository;
  readonly search: KnowledgeSearchRepository;
  readonly graphRepo: GraphRepository;
  readonly vectors: VectorRepository;
  readonly consolidation: ConsolidationRepository;
  readonly sessions: SessionRepository;
  readonly audit: AuditRepository;
  readonly graph: KnowledgeGraph;
  readonly coreMemory: CoreMemoryManager;
  readonly entities: EntityRepository;
  readonly conversations: ConversationRepository;

  private readonly _db: Database.Database;
  private currentSessionId: string | null = null;

  /** Expose raw DB for direct queries (used by UX routes). */
  get db(): Database.Database { return this._db; }

  constructor(db: Database.Database) {
    this._db = db;
    const memDb = new MemoryDatabaseManager(db);
    memDb.initialize();

    this.knowledge = new KnowledgeRepository(db);
    this.search = new KnowledgeSearchRepository(db);
    this.graphRepo = new GraphRepository(db);
    this.vectors = new VectorRepository(db);
    this.consolidation = new ConsolidationRepository(db);
    this.sessions = new SessionRepository(db);
    this.audit = new AuditRepository(db);
    this.graph = new KnowledgeGraph(this.graphRepo);
    this.graph.loadFromDb();
    this.coreMemory = new CoreMemoryManager(db);
    this.entities = new EntityRepository(db);
    this.conversations = new ConversationRepository(db);
  }

  /** Start a new session. */
  startSession(agentName?: string): string {
    const sid = this.sessions.startSession(agentName);
    this.currentSessionId = sid;
    this.audit.log('SESSION_START', undefined, sid);
    return sid;
  }

  /** End the current session. */
  endSession(): void {
    if (!this.currentSessionId) return;
    this.sessions.endSession(this.currentSessionId);
    this.audit.log('SESSION_END', undefined, this.currentSessionId);
    this.currentSessionId = null;
  }

  /** Get current session ID. */
  getSessionId(): string | null {
    return this.currentSessionId;
  }

  /** Get overall memory statistics. */
  getStats(): MemoryStats {
    const tierBreakdown = this.consolidation.getTierStats();
    const totalEntries = tierBreakdown.reduce((sum, t) => sum + t.entryCount, 0);
    return {
      totalEntries,
      totalEdges: this.graphRepo.countEdges(),
      totalVectors: this.vectors.count(),
      tierBreakdown,
    };
  }
}
